cli/paas/client.go

```diff
@@ -0,0 +1,389 @@
+func (c *CarrierClient) CreateOrg(org string) error {
+	_, _, err := c.giteaClient.CreateOrg(gitea.CreateOrgOption{
+		Name: org,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "failed to create org")
+	}
+
+	c.ui.Normal().WithStringValue("name", org).Msg("Organization created.")
+
+	return nil
+}
+
+// Delete deletes an app
+func (c *CarrierClient) Delete(app string) error {
+	return nil
+}
+
+// Orgs get a list of all orgs in gitea
+func (c *CarrierClient) Orgs() error {
+	orgs, _, err := c.giteaClient.AdminListOrgs(gitea.AdminListOrgsOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list orgs")
+	}
+
+	msg := c.ui.Normal().WithTable("Name")
+
+	for _, org := range orgs {
+		msg = msg.WithTableRow(org.UserName)
+	}
+
+	msg.Msg("Carrier Organizations")
+
+	return nil
+}
+
+// Push pushes an app
+func (c *CarrierClient) Push(app string, path string) error {
+	err := c.createRepo(app)
+	if err != nil {
+		return errors.Wrap(err, "create repo failed")
+	}
+
+	err = c.createRepoWebhook(app)
+	if err != nil {
+		return errors.Wrap(err, "webhook configuration failed")
+	}
+
+	tmpDir, err := c.prepareCode(app, path)
+	if err != nil {
+		return errors.Wrap(err, "failed to prepare code")
+	}
+
+	err = c.gitPush(app, tmpDir)
+	if err != nil {
+		return errors.Wrap(err, "failed to git push code")
+	}
+
+	stopFunc, err := c.logs(app)
+	if err != nil {
+		return errors.Wrap(err, "failed to tail logs")
+	}
+	defer stopFunc()
+
+	err = c.waitForApp(app)
+	if err != nil {
+		return errors.Wrap(err, "waiting for app failed")
+	}
+
+	return nil
+}
+
+// Target targets an org in gitea
+func (c *CarrierClient) Target(org string) error {
+	if org == "" {
+		c.ui.Normal().WithStringValue("Name", c.config.Org).Msg("Targetted org:")
+	}
+
+	c.config.Org = org
+	err := c.config.Save()
+	if err != nil {
+		return errors.Wrap(err, "failed to save configuration")
+	}
+
+	c.ui.Normal().WithStringValue("Name", c.config.Org).Msg("Targetted org:")
+
+	return nil
+}
+
+func (c *CarrierClient) check() {
+	c.giteaClient.GetMyUserInfo()
+}
+
+func (c *CarrierClient) createRepo(name string) error {
+	c.ui.Normal().Msg("Creating application ...")
+
+	_, _, err := c.giteaClient.CreateOrgRepo(c.config.Org, gitea.CreateRepoOption{
+		Name:          name,
+		AutoInit:      true,
+		Private:       true,
+		DefaultBranch: "main",
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "failed to create application")
+	}
+
+	c.ui.Normal().WithStringValue("name", name).Msg("Application created.")
+
+	return nil
+}
+
+func (c *CarrierClient) createRepoWebhook(name string) error {
+	c.ui.Normal().Msg("Creating webhook in the repo ...")
+
+	c.giteaClient.CreateRepoHook(c.config.Org, name, gitea.CreateHookOption{
+		Active:       true,
+		BranchFilter: "*",
+		Config: map[string]string{
+			"secret":       HookSecret,
+			"http_method":  "POST",
+			"url":          StagingEventListenerURL,
+			"content_type": "json",
+		},
+		Type: "gitea",
```

Comment:
What events is this hook triggering on ?
As I see no events listed at all I suspect this will not be triggered by anything.
Which would explain why nothing happens when I try to push the sample app.

Reply:
Hm. Adding 
> Events: []string{"push"},
did not help. Same nothingness and same timeout.

Note: Local cluster is `k3s` based. (In case this is different from the setup used by @viovanov ).

Reply:
Ok. I see that the event listener is triggered (not sure if because of the event decl I added, will be checked by removing.

```
{"level":"info","ts":"2021-01-18T11:22:38.972Z","logger":"eventlistener","caller":"sink/sink.go:237","msg":"ResolvedParams : [{Name:gitrevision Value:main} {Name:namespace Value:eirini-workloads} {Name:gitrepositoryurl Value:http://gitea.gitea:10080/workspace/sample4} {Name:appname Value:sample4}]","knative.dev/controller":"eventlistener","/triggers-eventid":"dhllt","/trigger":""}
{"level":"info","ts":"2021-01-18T11:22:38.974Z","logger":"eventlistener","caller":"resources/create.go:95","msg":"Generating resource: kind: &APIResource{Name:pipelineruns,Namespaced:true,Kind:PipelineRun,Verbs:[delete deletecollection get list patch create update watch],ShortNames:[pr prs],SingularName:pipelinerun,Categories:[tekton tekton-pipelines],Group:tekton.dev,Version:v1beta1,StorageVersionHash:RcAKAgPYYoo=,}, name: staging-pipeline-run-sample4-p5wps","knative.dev/controller":"eventlistener"}
{"level":"info","ts":"2021-01-18T11:22:38.974Z","logger":"eventlistener","caller":"resources/create.go:103","msg":"For event ID \"dhllt\" creating resource tekton.dev/v1beta1, Resource=pipelineruns","knative.dev/controller":"eventlistener"}
```

So there should be a `staging-pipeline-run-sample4-p5wps` somewhere. ... Seeing no pod however. ... Ah, a `pipelineruns` type ...

Look at that, all my attempts:

```
ork@tagetarl:~/SUSE/dev/carrier> k get pipelineruns :
kubectl get pipelineruns --all-namespaces
NAMESPACE          NAME                                 SUCCEEDED   REASON           STARTTIME   COMPLETIONTIME
eirini-workloads   staging-pipeline-run-sample-wzws7    False       CouldntGetTask   132m        132m
eirini-workloads   staging-pipeline-run-sample2-skg7m   False       CouldntGetTask   18m         18m
eirini-workloads   staging-pipeline-run-sample3-74fsc   False       CouldntGetTask   4m35s       4m35s
eirini-workloads   staging-pipeline-run-sample4-p5wps   False       CouldntGetTask   3m28s       3m28s
```

Even before I added "Events" to the structure, so that is indeed not required.
Retracting my change request here. Likely something in my setup now, not in the code.

Reply:
Getting more details on the object I find

```
status:
  completionTime: "2021-01-18T09:13:12Z"
  conditions:
  - lastTransitionTime: "2021-01-18T09:13:12Z"
    message: 'Pipeline eirini-workloads/staging-pipeline can''t be Run; it contains
      Tasks that don''t exist: Couldn''t retrieve Task "buildpacks": tasks.tekton.dev
      "buildpacks" not found'
    reason: CouldntGetTask
    status: "False"
```

Reply:
Now I wonder if the go-based tekton deployment is incomplete ? :bell: @jimmykarily

Reply:
Ok. We have tekton's `tasks` CRD.
```
work@tagetarl:~/SUSE/dev/carrier> k api-resources|grep tekton|grep tasks
kubectl api-resources
clustertasks                                   tekton.dev                     false        ClusterTask
tasks                                          tekton.dev                     true         Task
```
We have no actual tasks running however, and something else expects a `buildpacks` task to be present.
```
work@tagetarl:~/SUSE/dev/carrier> k get tasks :
kubectl get tasks --all-namespaces
No resources found.
```

Reply:
Ok, with the bash-based carrier the above commands return the same results. So that is not the exact difference.
That there is a difference between the go and bash tektons however is evident.
I am currently pushing to a cluster initialized from the bash scripts, and there pushing the sample-app does work.

I will see if I can dig out what the relevant differences are.

Reply:
That Task should be here: https://github.com/SUSE/carrier/blob/main/cli/embedded-files/tekton/staging.yaml#L4-L13